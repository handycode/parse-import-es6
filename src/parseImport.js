import strip from 'parse-comment-es6';
import parseImportClause from './parseImportClause';
import { trimWordSpacing, getAllLineStart, mapLocToRange } from './util';

// TODO: make line comment follow the last identifier
// TODO: extract all blockcomment (in new line or not decided by the occupied lines if the comment)
// TODO: we make the comment end of the line if exist the block comment in the begining of the line,
//       follow the first identifier of the line include 'import' 'from'
// TODO: handle there my be sentences between import statement and linecomment begin, low p
// TODO: handle there my be sentences between one line blockcomemnt and import statement, low p

const importRegex = /(?:import[\s]+)([\s\S]*?)(?:from[\s]+['|"](\w+)['|"](?:\s*;){0,1})/g;
/**
 * return all import statements
 * @param {*string} strippedText text without comments
 */
export function getAllImport(originText) {
    let res = null;
    const importList = [];
    const lineStart = getAllLineStart(originText);
    while ((res = importRegex.exec(originText)) != null) { // eslint-disable-line
        let importedDefaultBinding = null;
        let nameSpaceImport = null;
        let namedImports = null;
        let error = 0;
        const moduleSpecifier = res[2];
        const importClause = trimWordSpacing(res[1]);

        const parseResult = parseImportClause(importClause);
        if (parseResult != null) {
            importedDefaultBinding = parseResult.importedDefaultBinding;
            nameSpaceImport = parseResult.nameSpaceImport;
            namedImports = parseResult.namedImports;
        } else {
            error = 1;
        }

        importList.push({
            importedDefaultBinding,
            nameSpaceImport,
            namedImports,
            moduleSpecifier,
            range: {
                start: res.index,
                end: res.index + res[0].length,
            },
            loc: mapLocToRange(lineStart, res.index, res.index + res[0].length),
            raw: res[0],
            error,
        });
    }
    return importList;
}

// exculde the first leading comment of the first import, if exist 'flow' 'Copyright' 'LICENSE'
const ignoreComment = /@flow|license|copyright/i;

/**
 * return leading comment list
 * @param {*list<comment>} comments
 * @param {*} the last match leading comment of one import
 * @param {*} beginIndex the potential begin of the comment index of the import
 * @param {*} first whether first import
 */
function findLeadingComments(comments, index, beginIndex, first) {
    const leadComments = [];
    if (first && ignoreComment.test(comments[index].raw)) {
        return leadComments;
    }
    let backIndex = index - 1;
    while (backIndex >= beginIndex &&
        (comments[backIndex].loc.end.line + 1 === comments[backIndex + 1].loc.start.line
            || comments[backIndex].loc.end.line === comments[backIndex + 1].loc.start.line)) {
        if (first && ignoreComment.test(comments[backIndex].raw)) {
            break;
        }
        backIndex -= 1;
    }
    for (let ind = backIndex + 1; ind <= index; ind += 1) {
        leadComments.push(comments[ind]);
    }
    return leadComments;
}

function findTrailingComments(comments, index, nextImp) {
    const trailingComments = [];
    let forwardIndex = index;
    while (forwardIndex < comments.length - 1 &&
        (comments[forwardIndex].loc.end.line + 1 === comments[forwardIndex + 1].loc.start.line
            || comments[forwardIndex].loc.end.line === comments[forwardIndex + 1].loc.start.line)) {
        forwardIndex += 1;
    }
    for (let ind = index; ind <= forwardIndex; ind += 1) {
        trailingComments.push(comments[ind]);
        /**
         * check if the comment is next to the nextImp,
         * if true, put these comments into leading comments of next imp
         */
        if (nextImp && comments[ind].loc.end.line + 1 >= nextImp.loc.start.line) {
            return [];
        }
    }
    return trailingComments;
}

export function mapCommentsToImport(imp, beginIndex, comments = [], first = false, nextImp) {
    let leadComments = [];
    let trailingComments = [];
    let index;
    for (index = beginIndex; index < comments.length; index += 1) {
        const comment = comments[index];
        if (comment.loc.end.line + 1 === imp.loc.start.line
            // treat blockcomment before import(in same line) as leading comment
            // e.g /*asdfasdf
            //     asdf*/ import ...
            || (comment.loc.end.line === imp.loc.start.line && comment.loc.start.line < imp.loc.start.line)) {
            // look forward for the last match comment
            while (index + 1 < comments.length
                && comments[index + 1].loc.end.line + 1 === imp.loc.start.line) {
                index += 1;
            }
            leadComments = findLeadingComments(comments, index, beginIndex, first);
        }
        if (comment.loc.start.line === imp.loc.end.line + 1
            // treat blockcomment after import(in same line) as trailing comment
            // e.g import ... /*asdfasdf
            //     asdf*/
            || (comment.loc.start.line === imp.loc.end.line && comment.loc.end.line > imp.loc.end.line)) {
            trailingComments = findTrailingComments(comments, index, nextImp);
            // skip the trailingComments, there will make bug if multiple comments in same line
            if (trailingComments.length !== 0) {
                index += trailingComments.length - 1;
            }
        }
        /**
         * find interweave comment
         */
        if (comment.loc.start.line >= imp.loc.end.line + 1) {
            break;
        }
    }
    return [
        Object.assign({}, imp, {
            leadComments,
            trailingComments,
        }),
        index,
    ];
}

export default function parseImport(originText) {
    const imports = getAllImport(originText);
    const comments = strip(originText, { comment: true, range: true, loc: true, raw: true })
        .comments;

    const filterCommentImports = imports.filter((imp) => {
        // filter import in comment, TODO: filter all wapper statement like '/', '"', "`"
        for (let index = 0; index < comments.length; index += 1) {
            const comment = comments[index];
            if (comment.range.start <= imp.range.start && comment.range.end >= imp.range.end) {
                return false;
            }
        }
        return true;
    });

    const pickedImports = [];
    let commentIndex = 0;
    filterCommentImports.forEach((imp, index) => {
        const [res, rIndex] = mapCommentsToImport(
            imp, commentIndex, comments, index === 0, filterCommentImports[index + 1]);
        if (res != null) {
            commentIndex = rIndex;
            pickedImports.push(res);
        }
    });
    return pickedImports;
}

