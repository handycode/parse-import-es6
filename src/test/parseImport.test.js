import { getImportByRegex } from '../parseImport';

describe('test getImportByRegex', () => {
    test('get default correctly', () => {
        const p = "import a from 'aa'";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: 'a',
            bracketImport: [],
            importPath: 'aa',
            start: 0,
            end: 18,
            raw: "import a from 'aa'",
            error: 0,
        }]);
    });

    test('get named import correctly', () => {
        const p = "import { a } from 'aa'";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: null,
            bracketImport: ['a'],
            importPath: 'aa',
            start: 0,
            end: 22,
            raw: "import { a } from 'aa'",
            error: 0,
        }]);
    });

    test('get named import and default import correctly', () => {
        const p = "import b, { a } from 'aa'";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: 'b',
            bracketImport: ['a'],
            importPath: 'aa',
            start: 0,
            end: 25,
            raw: "import b, { a } from 'aa'",
            error: 0,
        }]);
    });

    test('get import correctly if named import is empty', () => {
        const p = "import b, { \r\n} from 'aa'";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: 'b',
            bracketImport: [],
            importPath: 'aa',
            start: 0,
            end: 25,
            raw: "import b, { \r\n} from 'aa'",
            error: 0,
        }]);
    });

    test('get import correctly if exist line feed', () => {
        const p = "import b\r, { \r\na \r\n, c} \n from 'aa';";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: 'b',
            bracketImport: ['a', 'c'],
            importPath: 'aa',
            start: 0,
            end: 36,
            raw: "import b\r, { \r\na \r\n, c} \n from 'aa';",
            error: 0,
        }]);
    });

    test('get import correctly if default import after named import', () => {
        const p = "import { \r\na \r\n, c}, b from 'aa';";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: 'b',
            bracketImport: ['a', 'c'],
            importPath: 'aa',
            start: 0,
            end: 33,
            raw: "import { \r\na \r\n, c}, b from 'aa';",
            error: 0,
        }]);
    });

    test('get import correctly if exist as', () => {
        const p = "import a as b, { c as d, f } from 'aa';";
        expect(getImportByRegex(p)).toEqual([{
            defaultImport: 'a as b',
            bracketImport: ['c as d', 'f'],
            importPath: 'aa',
            start: 0,
            end: 39,
            raw: "import a as b, { c as d, f } from 'aa';",
            error: 0,
        }]);
    });

    test('get error import if default import is error', () => {
        const p = "import a b, { c as d, f } from 'aa';";
        expect(getImportByRegex(p)).toEqual([{
            importPath: 'aa',
            start: 0,
            end: 36,
            raw: "import a b, { c as d, f } from 'aa';",
            error: 1,
        }]);
    });
});
