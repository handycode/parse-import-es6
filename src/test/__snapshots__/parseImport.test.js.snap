// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseImport close comment is treated as leading comment correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 2,
      },
      "start": Object {
        "column": 12,
        "line": 2,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 70,
      "start": 48,
    },
    "raw": "import { a } from 'aa'",
    "trailingComments": Array [],
  },
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 3,
          },
        },
        "range": Object {
          "end": 105,
          "start": 83,
        },
        "raw": "// i am a comment, two",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 4,
      },
      "start": Object {
        "column": 12,
        "line": 4,
      },
    },
    "moduleSpecifier": "bb",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
    ],
    "range": Object {
      "end": 140,
      "start": 118,
    },
    "raw": "import { b } from 'bb'",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 36,
            "line": 5,
          },
          "start": Object {
            "column": 12,
            "line": 5,
          },
        },
        "range": Object {
          "end": 177,
          "start": 153,
        },
        "raw": "// i am a comment, three",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport close comment will be treated as leading or trailing comment 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 15,
            "line": 4,
          },
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "range": Object {
          "end": 102,
          "start": 48,
        },
        "raw": "/**
             * i am a comment, two
             */",
        "type": "BlockComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 5,
      },
      "start": Object {
        "column": 12,
        "line": 5,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 137,
      "start": 115,
    },
    "raw": "import { a } from 'aa'",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 36,
            "line": 6,
          },
          "start": Object {
            "column": 12,
            "line": 6,
          },
        },
        "range": Object {
          "end": 174,
          "start": 150,
        },
        "raw": "// i am a comment, three",
        "type": "LineComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 35,
            "line": 7,
          },
          "start": Object {
            "column": 12,
            "line": 7,
          },
        },
        "range": Object {
          "end": 210,
          "start": 187,
        },
        "raw": "// i am a comment, four",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport filter commentted import 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 2,
      },
      "start": Object {
        "column": 12,
        "line": 2,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 70,
      "start": 48,
    },
    "raw": "import { a } from 'aa'",
    "trailingComments": Array [],
  },
]
`;

exports[`parseImport filter copyright 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 5,
          },
          "start": Object {
            "column": 12,
            "line": 5,
          },
        },
        "range": Object {
          "end": 155,
          "start": 133,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 6,
      },
      "start": Object {
        "column": 12,
        "line": 6,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 190,
      "start": 168,
    },
    "raw": "import { a } from 'aa'",
    "trailingComments": Array [],
  },
]
`;

exports[`parseImport filter flow 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 2,
          },
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "range": Object {
          "end": 55,
          "start": 33,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 3,
      },
      "start": Object {
        "column": 12,
        "line": 3,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 90,
      "start": 68,
    },
    "raw": "import { a } from 'aa'",
    "trailingComments": Array [],
  },
]
`;

exports[`parseImport parse leading comment in the same line correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": "a",
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 2,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 50,
          "start": 13,
        },
        "raw": "/*
            i am a comment, one */",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 56,
            "line": 2,
          },
          "start": Object {
            "column": 35,
            "line": 2,
          },
        },
        "range": Object {
          "end": 72,
          "start": 51,
        },
        "raw": "//i am a comment, two",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 46,
        "line": 3,
      },
      "start": Object {
        "column": 12,
        "line": 3,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
      "c as d",
    ],
    "range": Object {
      "end": 119,
      "start": 85,
    },
    "raw": "import a, { b, c as d } from 'aa';",
    "trailingComments": Array [],
  },
]
`;

exports[`parseImport parse multiple import correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 2,
      },
      "start": Object {
        "column": 12,
        "line": 2,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 70,
      "start": 48,
    },
    "raw": "import { a } from 'aa'",
    "trailingComments": Array [],
  },
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 4,
          },
          "start": Object {
            "column": 12,
            "line": 4,
          },
        },
        "range": Object {
          "end": 106,
          "start": 84,
        },
        "raw": "// i am a comment, two",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 34,
        "line": 5,
      },
      "start": Object {
        "column": 12,
        "line": 5,
      },
    },
    "moduleSpecifier": "bb",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
    ],
    "range": Object {
      "end": 141,
      "start": 119,
    },
    "raw": "import { b } from 'bb'",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 36,
            "line": 6,
          },
          "start": Object {
            "column": 12,
            "line": 6,
          },
        },
        "range": Object {
          "end": 178,
          "start": 154,
        },
        "raw": "// i am a comment, three",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport parse multiple leading comments in the same line correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": "a",
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 3,
          },
        },
        "range": Object {
          "end": 72,
          "start": 50,
        },
        "raw": "//i am a comment, four",
        "type": "LineComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 5,
          },
          "start": Object {
            "column": 12,
            "line": 4,
          },
        },
        "range": Object {
          "end": 122,
          "start": 85,
        },
        "raw": "/*
            i am a comment, one */",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 40,
            "line": 5,
          },
          "start": Object {
            "column": 35,
            "line": 5,
          },
        },
        "range": Object {
          "end": 128,
          "start": 123,
        },
        "raw": "/*a*/",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 46,
            "line": 5,
          },
          "start": Object {
            "column": 41,
            "line": 5,
          },
        },
        "range": Object {
          "end": 134,
          "start": 129,
        },
        "raw": "/*b*/",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 68,
            "line": 5,
          },
          "start": Object {
            "column": 47,
            "line": 5,
          },
        },
        "range": Object {
          "end": 156,
          "start": 135,
        },
        "raw": "//i am a comment, two",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 46,
        "line": 6,
      },
      "start": Object {
        "column": 12,
        "line": 6,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
      "c as d",
    ],
    "range": Object {
      "end": 203,
      "start": 169,
    },
    "raw": "import a, { b, c as d } from 'aa';",
    "trailingComments": Array [],
  },
]
`;

exports[`parseImport parse multiple trailing comments in the same line correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": "a",
    "leadComments": Array [],
    "loc": Object {
      "end": Object {
        "column": 46,
        "line": 1,
      },
      "start": Object {
        "column": 12,
        "line": 1,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
      "c as d",
    ],
    "range": Object {
      "end": 47,
      "start": 13,
    },
    "raw": "import a, { b, c as d } from 'aa';",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 17,
            "line": 2,
          },
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "range": Object {
          "end": 65,
          "start": 60,
        },
        "raw": "/*a*/",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 23,
            "line": 2,
          },
          "start": Object {
            "column": 18,
            "line": 2,
          },
        },
        "range": Object {
          "end": 71,
          "start": 66,
        },
        "raw": "/*b*/",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 15,
            "line": 3,
          },
          "start": Object {
            "column": 24,
            "line": 2,
          },
        },
        "range": Object {
          "end": 90,
          "start": 72,
        },
        "raw": "/*
            c*/",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 37,
            "line": 3,
          },
          "start": Object {
            "column": 16,
            "line": 3,
          },
        },
        "range": Object {
          "end": 112,
          "start": 91,
        },
        "raw": "//i am a comment, one",
        "type": "LineComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 33,
            "line": 4,
          },
          "start": Object {
            "column": 12,
            "line": 4,
          },
        },
        "range": Object {
          "end": 146,
          "start": 125,
        },
        "raw": "//i am a comment, two",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport parseImport complex import correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": "a",
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 46,
        "line": 2,
      },
      "start": Object {
        "column": 12,
        "line": 2,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
      "c as d",
    ],
    "range": Object {
      "end": 82,
      "start": 48,
    },
    "raw": "import a, { b, c as d } from 'aa';",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 3,
          },
        },
        "range": Object {
          "end": 117,
          "start": 95,
        },
        "raw": "// i am a comment, two",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport parseImport correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 35,
        "line": 2,
      },
      "start": Object {
        "column": 12,
        "line": 2,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 71,
      "start": 48,
    },
    "raw": "import { a } from 'aa';",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 3,
          },
        },
        "range": Object {
          "end": 106,
          "start": 84,
        },
        "raw": "// i am a comment, two",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport parseImport parse last comment in end line correctly 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": "a",
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 35,
          "start": 13,
        },
        "raw": "// i am a comment, one",
        "type": "LineComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 46,
        "line": 2,
      },
      "start": Object {
        "column": 12,
        "line": 2,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "b",
      "c as d",
    ],
    "range": Object {
      "end": 82,
      "start": 48,
    },
    "raw": "import a, { b, c as d } from 'aa';",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 34,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 3,
          },
        },
        "range": Object {
          "end": 117,
          "start": 95,
        },
        "raw": "// i am a comment, two",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport treat blockcomment after import(in same line) as trailing comment 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [],
    "loc": Object {
      "end": Object {
        "column": 35,
        "line": 1,
      },
      "start": Object {
        "column": 12,
        "line": 1,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 36,
      "start": 13,
    },
    "raw": "import { a } from 'aa';",
    "trailingComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 23,
            "line": 2,
          },
          "start": Object {
            "column": 36,
            "line": 1,
          },
        },
        "range": Object {
          "end": 71,
          "start": 37,
        },
        "raw": "/*asdfasdf
            asdfasdf */",
        "type": "BlockComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 26,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 3,
          },
        },
        "range": Object {
          "end": 98,
          "start": 84,
        },
        "raw": "//asdfasdfasdf",
        "type": "LineComment",
      },
    ],
  },
]
`;

exports[`parseImport treat blockcomment before import(in same line) as leading comment 1`] = `
Array [
  Object {
    "error": 0,
    "importedDefaultBinding": null,
    "leadComments": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 22,
            "line": 1,
          },
          "start": Object {
            "column": 12,
            "line": 1,
          },
        },
        "range": Object {
          "end": 23,
          "start": 13,
        },
        "raw": "//asdfasdf",
        "type": "LineComment",
      },
      Object {
        "loc": Object {
          "end": Object {
            "column": 22,
            "line": 3,
          },
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "range": Object {
          "end": 69,
          "start": 36,
        },
        "raw": "/*asdfasdf
            asdfasdf*/",
        "type": "BlockComment",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 45,
        "line": 3,
      },
      "start": Object {
        "column": 22,
        "line": 3,
      },
    },
    "moduleSpecifier": "aa",
    "nameSpaceImport": null,
    "namedImports": Array [
      "a",
    ],
    "range": Object {
      "end": 92,
      "start": 69,
    },
    "raw": "import { a } from 'aa';",
    "trailingComments": Array [],
  },
]
`;
